### PR TITLE
ColorPicker improvement

### PR DESCRIPTION
**Pull request by thangdinh**
The current color picker is hard to customize the style without copying and modifying the whole template file. This PR is to break down the color picker component into sub-components the color picker button (the one when you already selected a color) and the dropdown picker, which has a lot of color cells. 
Also, the current implementation always renders the dropdown and just hide it using css. It is not practical when we have to render a large amount of color pickers, for e.g. when we need to set colors for each item in a list. In this PR we only render the dropdown when we open it (click on the button).
Other things done: Refactor code, split files, replace view by component, give some love on the introduction site.
@Addepar/fire 
CSS/website change: @Addepar/design

### CODE REVIEW

File: src/color_picker.coffee [hunk 1]
```diff
@@ -1,258 +1,84 @@
...
-      @set 'selectedColor', color
-
-    sendCustomColor: ->
-      color = @get 'customColor'
+    ###*
+     * This action is bound to the colorPicker button to hide/show the dropdown
+     * when users click on it.
+    ###
+    toggleDropdown: ->
+      shouldOpenDropdown = @get 'shouldOpenDropdown'
```
Comment (philpee2) [inline src/color_picker.coffee]: Can this be `@toggleProperty('shouldOpenDropdown')`?

File: src/color_picker_cell.coffee [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+ * @class
+ * @augments {Ember.View, Ember.Widgets.StyleBindingMixin,
+ *            Ember.Widgets.ColorPickerMixin}
+###
+Ember.Widgets.ColorPickerCellComponent =
+Ember.Component.extend Ember.Widgets.StyleBindingsMixin,
+Ember.Widgets.ColorPickerMixin,
+  layoutName: 'color-picker-cell'
+  classNames: ['pull-left', 'color-picker-cell']
+  classNameBindings: Ember.A ['isActive:active:inactive']
```
Comment (philpee2) [inline src/color_picker_cell.coffee]: Does this need to be wrapped in `Ember.A`?

File: src/color_picker_cell.coffee [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+  styleBindings:  'color:background-color'
+
+  color: null
+
+  isActive: Ember.computed ->
+    @colorToHex(@get('controller.selectedColor')) is @colorToHex(@get('color'))
+  .property 'controller.selectedColor', 'color'
+
+  click: (event) ->
```
Comment (philpee2) [inline src/color_picker_cell.coffee]: `event` isn't being used

File: src/color_picker_dropdown.coffee [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+Ember.Component.extend Ember.Widgets.BodyEventListener,
+Ember.Widgets.ColorPickerMixin,
+  layoutName: 'color-picker-dropdown'
+  dropdownClass: null
+
+  selectedColor: ''
+
+  customColor: ''
+
+  colorRows:
```
Comment (philpee2) [inline src/color_picker_dropdown.coffee]: Should this be wrapped in `Ember.computed`?

File: src/color_picker.coffee [hunk 5]
```diff
@@ -1,258 +1,84 @@
...
+  ###
+  colorPickerButtonPartial: 'color-picker-button-partial'
+
+  ###*
+   * This is a boolean to control if we should render the colorPicker dropdown
+   * or not. Instead of hiding it using CSS, we use this flag to control the
+   * rendering.
+   * @type {Boolean}
+  ###
+  shouldOpenDropdown: false
```
Comment (philpee2) [inline src/color_picker.coffee]: I think something like `isDropdownOpen` is clearer

File: src/color_picker_cell.coffee (hunk 2 above)
Comment (cyril-sf) [inline src/color_picker_cell.coffee]: no

File: src/color_picker_dropdown.coffee (hunk 4 above)
Comment (cyril-sf) [inline src/color_picker_dropdown.coffee]: yes

File: src/color_picker_dropdown.coffee [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+        'transparent'
+      ])
+    ])
+
+  isColorTransparent: Ember.computed.equal 'selectedColorRGB', 'transparent'
+
+  setCustomColor: Ember.on 'init', Ember.observer ->
+    selectedColor = @get 'selectedColor'
+    selectedColor = @colorToHex(selectedColor)
+    return @set('customColor', '') if @get('colorRows').find (row) ->
```
Comment (cyril-sf) [inline src/color_picker_dropdown.coffee]: this looks weird to return a specific value in an observer.

Comment (philpee2) [inline src/color_picker_dropdown.coffee]: It's also weird to return the result of `set`, imo

**Comment by thangdinh** [comment]
Most of the comments are part of the old code, I rearrange them only. I will fix them still.

File: src/color_picker_dropdown.coffee [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+  formattedCustomColor: Ember.computed ->
+    customColor = @get('customColor').trim()
+    if customColor.charAt(0) isnt '#'
+      customColor = '#' + customColor
+    return customColor
+  .property 'customColor'
+
+  isCustomColorValid: Ember.computed ->
+    /(^#[0-9A-F]{6}$)|(^#[0-9A-F]{3}$)/i.test("#{@get('formattedCustomColor')}")
+  .property 'customColor'
```
Comment (PhyFrancis) [inline src/color_picker_dropdown.coffee]: `.property 'formattedCustomColor'`

File: src/color_picker_cell.coffee [hunk 8]
```diff
@@ -0,0 +1,25 @@
+###*
+ * ColorPickerCell
+ *
+ * This is an internal-use view to control the logic of a color cell in
+ * the color picker.
+ * @class
+ * @augments {Ember.View, Ember.Widgets.StyleBindingMixin,
+ *            Ember.Widgets.ColorPickerMixin}
+###
+Ember.Widgets.ColorPickerCellComponent =
```
Comment (PhyFrancis) [inline src/color_picker_cell.coffee]: it's good to also comment on what needs to be passed in, in order to use this component, e.g. `color`

File: src/color_picker.coffee [hunk 9]
```diff
@@ -1,258 +1,84 @@
...
-      return "transparent"
-    return rgbToHex(red, green, blue)
-  return undefined
-
-Ember.Widgets.ColorPickerComponent = Ember.Component.extend
+###*
+ * ColorPickerComponent
+ *
+ * This is the whole color picker component. It is the interface connecting with
+ * users.
```
Comment (embooglement) [inline src/color_picker.coffee]: This doc doesn't really make sense to me.

File: src/color_picker.coffee [hunk 10]
```diff
@@ -1,258 +1,84 @@
...
+ * @class
+ * @augments Ember.Component
```
Comment (embooglement) [inline src/color_picker.coffee]: Should also include `Ember.Widgets.ColorPickerMixin`

File: src/color_picker_dropdown.coffee [hunk 11]
```diff
@@ -0,0 +1,106 @@
...
+  ###
+  formattedCustomColor: Ember.computed ->
+    customColor = @get('customColor').trim()
+    if customColor.charAt(0) isnt '#'
+      customColor = '#' + customColor
+    return customColor
+  .property 'customColor'
+
+  isCustomColorValid: Ember.computed ->
+    /(^#[0-9A-F]{6}$)|(^#[0-9A-F]{3}$)/i.test("#{@get('formattedCustomColor')}")
```
Comment (embooglement) [inline src/color_picker_dropdown.coffee]: No need to use string interpolation on `@get('formattedCustomColor')`

File: src/color_picker_mixin.coffee [hunk 12]
```diff
@@ -0,0 +1,178 @@
...
+    violet: "#ee82ee"
+    wheat: "#f5deb3"
+    white: "#ffffff"
+    whitesmoke: "#f5f5f5"
+    yellow: "#ffff00"
+    yellowgreen: "#9acd32"
+
+  ###*
+   * Convert RGB color to Hex string
+   * @param  {Number} r Red channel
```
Comment (embooglement) [inline src/color_picker_mixin.coffee]: Probably handy to specify that the numbers have to be between 0 and 255.

File: tests/unit/color_picker_test.coffee [hunk 13]
```diff
@@ -11,65 +29,119 @@ testHexConversion = (colorPicker, color, hex) ->
 
 test 'Color picker converts color to hex when color is undefined', ->
   expect 3
+  colorPicker = @subject()
   [undefined, null, 0].forEach (color) =>
```
Comment (embooglement) [inline tests/unit/color_picker_test.coffee]: This doesn't need to be a fat arrow function anymore.

File: src/color_picker.coffee (hunk 1 above)
Comment (thangdinh) [inline src/color_picker.coffee]: Right. Forgot about that function. Thanks.

File: src/color_picker_cell.coffee (hunk 2 above)
Comment (thangdinh) [inline src/color_picker_cell.coffee]: It existed there before. Changed it.

File: src/color_picker_cell.coffee (hunk 3 above)
Comment (thangdinh) [inline src/color_picker_cell.coffee]: Existing code. Changed it.

File: src/color_picker_dropdown.coffee (hunk 4 above)
Comment (thangdinh) [inline src/color_picker_dropdown.coffee]: Existing code. I didn't intend to rewrite everything :D

File: src/color_picker.coffee (hunk 5 above)
Comment (thangdinh) [inline src/color_picker.coffee]: isDropdownOpen sounds like a check, shouldOpenDropdown is an order (for rendering). But I don't have strong opinion here.

File: src/color_picker_dropdown.coffee (hunk 6 above)
Comment (thangdinh) [inline src/color_picker_dropdown.coffee]: This was there before. I agree that it's weird here, but I would prefer to spend time removing this entire observer separately.

File: src/color_picker_dropdown.coffee (hunk 7 above)
Comment (thangdinh) [inline src/color_picker_dropdown.coffee]: Thanks. Nice catch!

File: src/color_picker.coffee (hunk 9 above)
Comment (thangdinh) [inline src/color_picker.coffee]: Any suggestion on this?

File: src/color_picker.coffee (hunk 10 above)
Comment (thangdinh) [inline src/color_picker.coffee]: Sure. I forgot to update the doc after adding this mixin.

File: src/color_picker_dropdown.coffee (hunk 11 above)
Comment (thangdinh) [inline src/color_picker_dropdown.coffee]: Done

File: src/color_picker_mixin.coffee (hunk 12 above)
Comment (thangdinh) [inline src/color_picker_mixin.coffee]: Done.

File: tests/unit/color_picker_test.coffee (hunk 13 above)
Comment (thangdinh) [inline tests/unit/color_picker_test.coffee]: Done

File: src/color_picker.coffee (hunk 9 above)
Comment (embooglement) [inline src/color_picker.coffee]: So this is the sentence that doesn't make sense to me: `It is the interface connecting with users`. What are you trying to communicate here?

Comment (thangdinh) [inline src/color_picker.coffee]: I think it wasn't good at the beginning. I would like to emphasize that it is the component but it does have sub-components where most of the logics are in the dropdown component.

File: src/color_picker_cell.coffee (hunk 8 above)
Comment (thangdinh) [inline src/color_picker_cell.coffee]: I think we can pass in any attributes belonging to the component, right? I guess we should put comment on the attributes inside the component. This code was moved from existing code (splitting the file) that's why there is no comment here.

File: app/templates/ember_widgets/color_picker.hbs [hunk 14]
```diff
@@ -6,24 +6,46 @@
...
+          &#123;&#123;color-picker
+            class="btn btn-default"
+            dropdownClass="pull-left"
+          &#125;&#125;</pre>
+      </div>
+    </div>
 
     <div class="col-md-3">
-      {{view Ember.Widgets.ColorPicker class="btn btn-default" dropdownClass="pull-right"}}
+      {{color-picker class="btn btn-default" dropdownClass="pull-right"}}
```
Comment (cyril-sf) [inline app/templates/ember_widgets/color_picker.hbs]: is there a reason why you let the `class` in that case and not in the previous example?

File: app/templates/ember_widgets/color_picker.hbs [hunk 15]
```diff
@@ -6,24 +6,46 @@
...
     <div class="col-md-3">
-      {{view Ember.Widgets.ColorPicker class="btn btn-default"}}
+      {{color-picker}}
       <button type="button" class="btn btn-default">Comparison</button>
     </div>
+    <div class="col-md-9">
+      <div class="highlight">
+        <pre class="prettyprint lang-html">
+          &#123;&#123;color-picker
+            class="btn btn-default"
```
Comment (cyril-sf) [inline app/templates/ember_widgets/color_picker.hbs]: those are not the parameters passed to the example.

File: app/templates/ember_widgets/color_picker.hbs (hunk 14 above)
Comment (thangdinh) [inline app/templates/ember_widgets/color_picker.hbs]: Now looking back these "examples" I don't think it make any sense with the button Comparison next to it. I think what we wanted before is: "Hey that's what it looks like, this is what the default button looks like". Right now the class doesn't do anything, I think.

**Comment by thangdinh** [comment]
Do we have any other comment?

File: src/templates/color-picker-button-partial.hbs [hunk 16]
```diff
@@ -0,0 +1,14 @@
...
+  {{#if isColorTransparent}}
+    {{color-picker-cell
+      color=selectedColor
+      classNames="color-picker-cell-transparent"
+    }}
+  {{else}}
+    {{color-picker-cell
+      color=selectedColor
+    }}
+  {{/if}}
```
Comment (embooglement) [inline src/templates/color-picker-button-partial.hbs]: Can't this be 
```
{{color-picker-cell
  color=selectedColor
  classNames="isColorTransparent:color-picker-cell-transparent"
}}
```

**Comment by embooglement** [comment]
LGTM

File: src/templates/color-picker-button-partial.hbs (hunk 16 above)
Comment (thangdinh) [inline src/templates/color-picker-button-partial.hbs]: Good catch! I just moved around the content didn't pay attention to this.

**Pull request merged**